// Copyright 2020 The 9nFL Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

//
// Fetch datablock from DataCenter
//

package jdfl;

enum StatusCode {
  OK = 0;
  NOT_READY = 1;
  FINISHED = 2;
  NOT_FOUND = 3;
  ERROR_ABORTED = 4;
}

message FetchDataBlockRequest {
  string request_id = 1;
};

message DataBlockMeta {
   string block_id = 1;
   int32 partition_id = 2; 
   int32 file_version = 3;
   int64 start_time = 4;
   int64 end_time = 5;
   repeated bytes example_ids = 6;
   int64 leader_start_index = 7;
   int64 leader_end_index = 8;
   int64 follower_start_index = 9;
   int64 follower_end_index = 10;
   int64 data_block_index = 11;
}

message DataBlockInfo {
  string request_id = 1;
  string data_path = 2;
  DataBlockMeta meta_info = 3; 
}

message FetchDataBlockResponse {
  StatusCode status_code = 1;
  string status_error_message = 2;
  DataBlockInfo db_info = 3;
}

service DataBlockQueryService {
  rpc QueryDataBlock (FetchDataBlockRequest) returns (FetchDataBlockResponse) {}
}

